### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by KubaKlin**

### CODE REVIEW

**Comment by mwanago** [comment]
Handle long words properly
Screenshot from 2024-11-19 02-51-34

File: index.html [hunk 1]
```diff
@@ -3,8 +3,25 @@
 <head>
   <meta charset="UTF-8">
   <title>Title</title>
+  <link rel="stylesheet" href="style.css">
 </head>
 <body>
```
Comment (mwanago) [inline index.html]: Fix your indentation

File: index.js [hunk 2]
```diff
@@ -1 +1,70 @@
...
+const tasksWrapper = document.querySelector('.tasks-column');
+const errorLabel = document.querySelector('.add-input-error-label');
+errorLabel.innerText = 'Task title cannot be empty';
+
+const firstColumn = document.querySelector('#first-column');
+const secondColumn = document.querySelector('#second-column');
+const thirdColumn = document.querySelector('#third-column');
+
+
+taskInputButton.addEventListener('click', function() {
```
Comment (mwanago) [inline index.js]: What if `taskInputButton` is null?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: index.js [hunk 3]
```diff
@@ -1 +1,70 @@
...
+  let taskTitle = taskInput.value;
```
Comment (mwanago) [inline index.js]: What if `taskInput` is null?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: index.js [hunk 4]
```diff
@@ -1 +1,70 @@
...
+      firstColumn.append(taskCard);
+      columnCount -= 1;
+    } else if (columnCount === 3) {
+      secondColumn.append(taskCard);
+      columnCount -= 1;
+    }
+  })
+
+  if (taskTitle !== '') {
+    tasksWrapper.append(taskCard);
```
Comment (mwanago) [inline index.js]: What if `tasksWrapper` is null?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: index.js [hunk 5]
```diff
@@ -1 +1,70 @@
...
+    taskInput.value = '';
+    errorLabel.classList.remove('active');
```
Comment (mwanago) [inline index.js]: What if `errorLabel` is null?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: index.js [hunk 6]
```diff
@@ -1 +1,70 @@
...
+      thirdColumn.append(taskCard);
+      columnCount += 1;
+    }
+  })
+
+  leftArrowButton.addEventListener('click', function() {
+    taskCard.cloneNode(true);
+    taskCard.remove();
+    if (columnCount === 2) {
+      firstColumn.append(taskCard);
```
Comment (mwanago) [inline index.js]: What if `firstColumn` is null?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: index.js [hunk 7]
```diff
@@ -1 +1,70 @@
...
+      columnCount -= 1;
+    } else if (columnCount === 3) {
+      secondColumn.append(taskCard);
```
Comment (mwanago) [inline index.js]: What if `secondColumn` is null?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: index.js [hunk 8]
```diff
@@ -1 +1,70 @@
...
+
+  let columnCount = 1;
+  rightArrowButton.addEventListener('click', function() {
+    taskCard.cloneNode(true);
+    taskCard.remove();
+    if (columnCount === 1) {
+      secondColumn.append(taskCard);
+      columnCount += 1;
+    } else if (columnCount === 2) {
+      thirdColumn.append(taskCard);
```
Comment (mwanago) [inline index.js]: What if `thirdColumn` is null?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: index.js (hunk 3 above)
Comment (mwanago) [inline index.js]: There is no need to use `let` here

File: index.js (hunk 2 above)
Comment (mwanago) [inline index.js]: This function is very, very long. Split it into multiple separate functions, such as `getButtonsWrapper` that returns the `buttonsWrapper` with all of the buttons.

```
taskInputButton.addEventListener('click', function() {
  // ...
  const taskCard = document.createElement('div');
  const buttonsWrapper = getButtonsWrapper();
  taskCard.append(buttonsWrapper);
  // ...
}
```

File: index.js [hunk 9]
```diff
@@ -1 +1,70 @@
...
+  rightArrowButton.classList.add('button-right-arrow');
+  buttonsWrapper.append(rightArrowButton);
+
+  binButton.addEventListener('click', function() {
+    taskCard.remove();
+  });
+
+  let columnCount = 1;
+  rightArrowButton.addEventListener('click', function() {
+    taskCard.cloneNode(true);
```
Comment (mwanago) [inline index.js]: Why `cloneNode()` before `remove()`?

File: index.js [hunk 10]
```diff
@@ -1 +1,70 @@
...
+
+  const rightArrowButton = document.createElement('button');
+  rightArrowButton.classList.add('button-right-arrow');
+  buttonsWrapper.append(rightArrowButton);
+
+  binButton.addEventListener('click', function() {
+    taskCard.remove();
+  });
+
+  let columnCount = 1;
```
Comment (mwanago) [inline index.js]: Cool idea, but maybe the name `currentColumnNumber` would be better

File: style.css [hunk 11]
```diff
@@ -0,0 +1,143 @@
+* {
+  margin: 0;
+  padding: 0;
+  box-sizing: border-box;
+}
+
+body {
+  display: flex;
+  min-height: 100vh;
```
Comment (mwanago) [inline style.css]: Use percentages instead

File: style.css [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+  background: #c9c9c9;
+  justify-content: center;
+  flex-direction: column;
+  align-items: center;
+  font-weight: 400;
+  font-family: Arial, Helvetica, sans-serif;
+}
+
+.add-task-wrapper {
+  position: relative;
```
Comment (mwanago) [inline style.css]: Why `position: relative`?

File: style.css [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+  display: flex;
+  align-items: center;
+  justify-content: center;
+}
+
+.add-input-error-label {
+  color: #fff;
+  position: absolute;
```
Comment (mwanago) [inline style.css]: Don't use `position: absolute` if not absolutely necessary. It makes the layout more complicated and unpredictable.

File: style.css [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+.add-task-button:hover {
+  background-color: darkgreen;
+  cursor: pointer;
+}
+
+.columns-wrapper {
+  width: 100%;
+  height: 75vh;
```
Comment (mwanago) [inline style.css]: Use percentages instead

File: style.css [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+  justify-content: space-around;
+  margin-top: 50px;
+}
+
+.tasks-column {
+  padding: 10px;
+  background-color: #fff;
+  width: 30%;
+  border-radius: 5px;
+  overflow: hidden;
```
Comment (mwanago) [inline style.css]: Why `overflow: hidden;`?

File: style.css [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+.tasks-column:first-of-type .button-left-arrow {
+  opacity: 0;
+  pointer-events: none;
+}
+.tasks-column:last-of-type .button-right-arrow {
+  opacity: 0;
+  pointer-events: none;
+}
```
Comment (mwanago) [inline style.css]: This is a little hacky. Personally, I'd do that through JavaScript to be more explicit.

**Review by mwanago — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (KubaKlin) [reply index.js]: this is the same method to check if element is present as typing 'if' with all elements like we used to do in previous excercises?

File: index.js (hunk 9 above)
Comment (KubaKlin) [reply index.js]: there will be nothing to clone if I delete it in the first place?

File: style.css (hunk 15 above)
Comment (KubaKlin) [reply style.css]: to proper work of border-radius but now it is redundant

File: index.js (hunk 2 above)
Comment (mwanago) [reply index.js]: Doing
```
if (taskInputButton) {
  taskInputButton.addEventListener(...)
}
```
and
```
taskInputButton?.addEventListener(...)
```
is the same. It was introduced to JavaScript around the year 2020. Very useful.

File: index.js [hunk 17]
```diff
@@ -1 +1,90 @@
...
+  })
+
+  leftArrowButton.addEventListener('click', function() {
+    moveTaskBackward(taskCard, currentColumnNumber, leftArrowButton, rightArrowButton);
+    currentColumnNumber -= 1;
+  })
+
+  if (taskTitle !== '') {
+    tasksWrapper?.append(taskCard);
+    taskInput.value = '';
```
Comment (mwanago) [inline index.js]: What if `taskInput` is `null`?

File: index.js [hunk 18]
```diff
@@ -1 +1,90 @@
-console.log('Happy developing ✨')
+const taskInput = document.querySelector('.add-task-input');
+const taskInputButton = document.querySelector('.add-task-button');
+const tasksWrapper = document.querySelector('.tasks-column');
+const errorLabel = document.querySelector('.add-input-error-label');
+errorLabel.innerText = 'Task title cannot be empty';
```
Comment (mwanago) [inline index.js]: What if `errorLabel` is null?

**Review by mwanago — APPROVED** [review]

File: style.css (hunk 12 above)
Comment (KubaKlin) [reply style.css]: redundant - removed

**Pull request merged**